### PR TITLE
[SYCLomatic] Add specialization of sycl::is_device_copyable for non-trivially-copyable user-defined type if passed to device side

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 1]
```diff
@@ -8639,6 +8729,15 @@ void KernelCallRule::runRule(
     if (DpctGlobalInfo::getInstance().buildLaunchKernelInfo(LaunchKernelCall)) {
       emplaceTransformation(new ReplaceStmt(LaunchKernelCall, true, false, ""));
       removeTrailingSemicolon(LaunchKernelCall, Result);
+      if (const DeclRefExpr *DRE =
+              getAddressedRef(LaunchKernelCall->getArg(0))) {
+        if (const Decl *D = DRE->getReferencedDeclOfCallee()) {
+          if (const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(D)) {
+            for (const ParmVarDecl *Parm : FD->parameters())
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]: param's type may different from call arg's type.

File: clang/test/dpct/kernel_without_name.cu [hunk 2]
```diff
@@ -277,3 +277,28 @@ void run_foo7(T *a, const T *b, const T *c, const T *d, const T *e, const int f,
...
+
+struct UserStruct2 {
+  UserStruct2() {}
+  UserStruct2(UserStruct2 const &) {}
+};
+// CHECK: template<>
+// CHECK-NEXT: struct sycl::is_device_copyable<UserStruct2> : std::true_type {};
+
+template<class V1, class V2>
+__global__ void foo_kernel7(UserStruct1<V1, V2>, UserStruct2) {}
```
Comment (tangjj11) [inline clang/test/dpct/kernel_without_name.cu]: should we test nested type like: `UserStruct1<UserStruct2, V2>`

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 3]
```diff
@@ -8563,6 +8565,91 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+      TArgs += ", ";
+    }
+    if (!TArgs.empty()) {
+      TArgs = TArgs.substr(0, TArgs.size() - 2);
+    }
+    OS << TD->getNameAsString() << "<" << TArgs << ">";
+    OS << "> : std::true_type {};";
+    OS << getNL();
+  } else if (const auto *RD = dyn_cast<RecordDecl>(D)) {
+    Repl += "template<>";
```
Comment (tangjj11) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
    Repl += "template <>";
```

**Review by tangjj11 — APPROVED** [review]

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: CUDA just do plain memcpy for the kernel arguments.
So, I think using the parameter type is as same as the argument type (ignoring the implicit case).

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: Fixed

File: clang/test/dpct/kernel_without_name.cu (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/kernel_without_name.cu]: test added

File: clang/test/dpct/is_dev_copyable_spec.cu [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+  UserStruct6(UserStruct6 const &) {}
+};
+
+template<class V>
+__global__ void k3(V) {}
+
+void foo3() {
+  UserStruct6<UserStruct5> us6;
+  k3<<<1, 1>>>(us6);
+}
```
Comment (zhiweij1) [inline clang/test/dpct/is_dev_copyable_spec.cu:119]: 1. Make foo2 template, and instantiate it later.
2. Forward decl
3. Specialization
4. Partial Specialization
5. Nested struct (public member)
6. something like matrix_info_t

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 5]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+  // Find insert location (next line after semicolon)
+  SourceLocation EndLoc = D->getEndLoc();
+  bool MeetSemicolon = false;
+  const char *CharPtr =
+      DpctGlobalInfo::getSourceManager().getCharacterData(EndLoc);
+  while (CharPtr) {
+    if (*CharPtr == ';') {
+      MeetSemicolon = true;
+      CharPtr++;
+      continue;
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 6]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+    }
+    if (*CharPtr == '\n') {
+      CharPtr++;
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 7]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+      if (MeetSemicolon)
+        break;
+      continue;
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 8]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+    }
+    CharPtr++;
+  }
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: Can use Lexer here?

**Review by ziranzha — COMMENTED** [review]
Try to confirm such cases when argument's type is `A::B` or `A::C`:
```
class A {
    using B = E;
    template<class T1, class T2> using C = D<T1, T2>;
};
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 9]
```diff
@@ -8641,6 +8733,15 @@ void KernelCallRule::runRule(
     if (DpctGlobalInfo::getInstance().buildLaunchKernelInfo(LaunchKernelCall)) {
       emplaceTransformation(new ReplaceStmt(LaunchKernelCall, true, false, ""));
       removeTrailingSemicolon(LaunchKernelCall, Result);
+      if (const DeclRefExpr *DRE =
+              getAddressedRef(LaunchKernelCall->getArg(0))) {
+        if (const Decl *D = DRE->getReferencedDeclOfCallee()) {
+          if (const FunctionDecl *FD = dyn_cast_or_null<FunctionDecl>(D)) {
+            for (const ParmVarDecl *Parm : FD->parameters())
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: Duplicated with code in AnalysisInfo.cpp

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 10]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+  } else {
+    return;
+  }
+
+  // Find insert location (next line after semicolon)
+  SourceLocation EndLoc = D->getEndLoc();
+  bool MeetSemicolon = false;
+  const char *CharPtr =
+      DpctGlobalInfo::getSourceManager().getCharacterData(EndLoc);
+  while (CharPtr) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]:
```suggestion
  while (CharPtr[Offset]) {
```

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 11]
```diff
@@ -8565,6 +8565,95 @@ if (CodePinInstrumentation.find(KCallSpellingRange.first) !=
...
+    Repl += "struct sycl::is_device_copyable<";
+    Repl += RD->getNameAsString();
+    Repl += "> : std::true_type {};";
+    Repl += getNL();
+  } else {
+    return;
+  }
+
+  // Find insert location (next line after semicolon)
+  SourceLocation EndLoc = D->getEndLoc();
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: Emit a warning for macro case and exit

File: clang/test/dpct/is_dev_copyable_spec.cu [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+// CHECK-NEXT:   UserStruct3(UserStruct3 const &) {}
+// CHECK-NEXT: };
+// CHECK-NEXT: template <class T1, class T2>
+// CHECK-NEXT: struct sycl::is_device_copyable<UserStruct3<T1, T2>> : std::true_type {};
+// CHECK-NEXT: struct UserStruct4 {
+// CHECK-NEXT:   UserStruct4() {}
+// CHECK-NEXT:   UserStruct4(UserStruct4 const &) {}
+// CHECK-NEXT: };
+// CHECK-NEXT: template <>
+// CHECK-NEXT: struct sycl::is_device_copyable<UserStruct4> : std::true_type {};
```
Comment (zhiweij1) [inline clang/test/dpct/is_dev_copyable_spec.cu:66]:
```suggestion
// CHECK-NEXT: struct ::sycl::is_device_copyable<UserStruct4> : std::true_type {};
```

Comment (zhiweij1) [reply clang/test/dpct/is_dev_copyable_spec.cu:66]: ::sycl::is_device_copyable cannot resolve the issue that the user's class is defined in a namespace.

File: clang/test/dpct/is_dev_copyable_spec.cu (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/is_dev_copyable_spec.cu:119]: 1,2,3,4 are doable.
5 and 6 need further discussion.

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 8 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: Done

**Comment by zhiweij1** [comment]
> class A {
>     using B = E;
>     template<class T1, class T2> using C = D<T1, T2>;
> };

Doable

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 13]
```diff
@@ -8568,6 +8568,17 @@ void KernelCallRule::runRule(
     if (Flag)
       DpctGlobalInfo::insertKCIndentWidth(IndentLen);
 
+    for (const Expr *Arg : KCall->arguments()) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:8579]: For the type we added specialization, we need to insert a warning at the kernel call location.
For all types not trivially copyable, we need to emit warning more accurately (direct base type/field/copy ctor/copy assignment/move ctor/move assignment/dtor)

File: clang/lib/DPCT/Utility.cpp [hunk 14]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+      return getTypeDecl(QT.getTypePtr()
+                             ->castAs<SubstTemplateTypeParmType>()
+                             ->getReplacementType());
+    default:
+      return nullptr;
+    }
+  };
+
+  std::function<void(QualType)> findNonTrivalCopyableType;
+  findNonTrivalCopyableType = [&findNonTrivalCopyableType,
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: move out to be a function

File: clang/lib/DPCT/Utility.cpp [hunk 15]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+        if (!isUserDefinedDecl(ClassDecl))
+          return;
+        std::vector<std::string> Messages;
+        if (!ClassDecl->hasSimpleCopyConstructor()) {
+          Messages.push_back("copy constructor");
+        }
+        if (!ClassDecl->hasSimpleCopyAssignment()) {
+          Messages.push_back("copy assignment");
+        }
+        if (!ClassDecl->hasSimpleMoveConstructor() &&
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: try to use needsimplicitmoveconstructor.

File: clang/lib/DPCT/Utility.cpp [hunk 16]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    if (CanonicalType->isArrayType()) {
+      return findNonTrivalCopyableType(
+          DpctGlobalInfo::getContext().getBaseElementType(CanonicalType));
+    }
+    if (const auto *RT = CanonicalType->getAs<RecordType>()) {
+      if (const auto *ClassDecl = dyn_cast<CXXRecordDecl>(RT->getDecl())) {
+        if (!isUserDefinedDecl(ClassDecl))
+          return;
+        std::vector<std::string> Messages;
+        if (!ClassDecl->hasSimpleCopyConstructor()) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:5016]: copy/move ctor/assign, delete 3 and leave 1 is also legal.

File: clang/lib/DPCT/Utility.cpp [hunk 17]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+                               "\"");
+          }
+        }
+        for (const auto &B : ClassDecl->bases()) {
+          bool FurtherAnalyze = false;
+          if (B.isVirtual()) {
+            Messages.push_back(
+                "virtual base class \"" +
+                DpctGlobalInfo::getOriginalTypeName(B.getType()) + "\"");
+            FurtherAnalyze = true;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
```

File: clang/lib/DPCT/Utility.cpp [hunk 18]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+          }
+          if (!B.getType().isTriviallyCopyableType(
+                  DpctGlobalInfo::getContext())) {
+            Messages.push_back(
+                "non trivially copyable base class \"" +
+                DpctGlobalInfo::getOriginalTypeName(B.getType()) + "\"");
+            FurtherAnalyze = true;
+          }
+          if (FurtherAnalyze)
+            findNonTrivalCopyableType(B.getType());
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: move to 5058

File: clang/lib/DPCT/Utility.cpp [hunk 19]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    case Type::TypeClass::SubstTemplateTypeParm:
+      return getTypeDecl(QT.getTypePtr()
+                             ->castAs<SubstTemplateTypeParmType>()
+                             ->getReplacementType());
+    default:
+      return nullptr;
+    }
+  };
+
+  std::function<void(QualType)> findNonTrivalCopyableType;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
  std::function<void(QualType)> checkTrivallyCopyable;
```

File: clang/lib/DPCT/Utility.cpp [hunk 20]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+          MsgStr = Messages[0] + " and " + Messages[1];
+        } else {
+          for (size_t i = 0; i < Messages.size() - 2; i++) {
+            MsgStr += (Messages[i] + ", ");
+          }
+          MsgStr += (Messages[Messages.size() - 2] + " and " +
+                     Messages[Messages.size() - 1]);
+        }
+        ReportLocations.insert(
+            std::make_pair(ClassDecl->getBeginLoc(), MsgStr));
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: pass a rule pointer here to report directly.

File: clang/lib/DPCT/Utility.cpp [hunk 21]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    Repl += getNL();
+    Repl += "struct sycl::is_device_copyable<";
+    Repl += NamespaceStr;
+    Repl += RD->getNameAsString();
+    Repl += "> : std::true_type {};";
+  } else {
+    return std::nullopt;
+  }
+
+  return std::make_pair(InsertLoc, Repl);
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: insert code can be moved into this function, then the return type can just be bool

File: clang/lib/DPCT/Utility.cpp [hunk 22]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    return std::nullopt;
+  if (Type.isTriviallyCopyableType(DpctGlobalInfo::getContext()))
+    return std::nullopt;
+
+  findNonTrivalCopyableType(Type);
+
+  std::function<const NamespaceDecl *(const DeclContext *, std::string &)>
+      getOutermostNamespaceDecl;
+  const NamespaceDecl *NS = nullptr;
+  getOutermostNamespaceDecl = [&getOutermostNamespaceDecl,
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: can use a while to get the same functionality

File: clang/lib/DPCT/Utility.cpp [hunk 23]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+      return nullptr;
+    }
+  };
+
+  // Find insert location
+  auto getInsertLocation = [](SourceLocation InsertLoc, tok::TokenKind TK) {
+    Token Tok;
+    std::optional<Token> OptTok = std::nullopt;
+    Lexer::getRawToken(InsertLoc, Tok, DpctGlobalInfo::getSourceManager(),
+                       DpctGlobalInfo::getContext().getLangOpts());
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: auto SM=...
auto Ctx = ...

File: clang/lib/DPCT/Utility.cpp [hunk 24]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    if (Tok.is(TK)) {
+      InsertLoc = Tok.getEndLoc();
+    } else {
+      while (OptTok = Lexer::findNextToken(
+                 InsertLoc, DpctGlobalInfo::getSourceManager(),
+                 DpctGlobalInfo::getContext().getLangOpts())) {
+        InsertLoc = OptTok.value().getEndLoc();
+        if (OptTok.value().is(TK)) {
+          break;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:5141]: add an assert to check if opttok is empty

File: clang/lib/DPCT/Utility.cpp [hunk 25]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+        Lexer::MeasureTokenLength(InsertLoc, DpctGlobalInfo::getSourceManager(),
+                                  DpctGlobalInfo::getContext().getLangOpts()));
+    return InsertLoc;
+  };
+
+  std::string NamespaceStr;
+  SourceLocation InsertLoc;
+  if (D->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit) {
+    if (D->getEndLoc().isMacroID())
+      return std::nullopt;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: cannot distinguish "cannot analyze" or "is copyable"

File: clang/lib/DPCT/Utility.cpp [hunk 26]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+    InsertLoc = getInsertLocation(D->getEndLoc(), tok::TokenKind::semi);
+  } else if (const auto *ND =
+                 getOutermostNamespaceDecl(D->getDeclContext(), NamespaceStr)) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: inline getOutermostNamespaceDecl here

File: clang/lib/DPCT/Utility.cpp [hunk 27]
```diff
@@ -4953,5 +4956,242 @@ int isArgumentInitialized(
...
+  } else if (auto UO = dyn_cast<UnaryOperator>(E)) {
+    if (UO->getOpcode() == UO_AddrOf) {
+      return getAddressedRef(UO->getSubExpr());
+    }
+  }
+  return nullptr;
+}
+
+std::optional<std::pair<SourceLocation, std::string>>
+analyzeDeviceCopyable(QualType Type,
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: split the 2 functionality into 2 functions.

File: clang/lib/DPCT/Utility.cpp [hunk 28]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+            Messages.push_back("move constructor");
+          } else if (M->isMoveAssignmentOperator()) {
+            Messages.push_back("move assignment");
+          }
+        }
+      }
+      if (!ClassDecl->hasSimpleDestructor()) {
+        Messages.push_back("destructor");
+      }
+      for (const auto &M : ClassDecl->methods()) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: merge 5001 and 5013

File: clang/lib/DPCT/Utility.cpp [hunk 29]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+      }
+      std::string MsgStr;
+      if (Messages.empty()) {
+        assert(0 && "Messages is empty.");
+      } else if (Messages.size() == 1) {
+        MsgStr = Messages[0];
+      } else if (Messages.size() == 2) {
+        MsgStr = Messages[0] + " and " + Messages[1];
+      } else {
+        for (size_t i = 0; i < Messages.size() - 2; i++) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: Use iterator

File: clang/lib/DPCT/Utility.cpp [hunk 30]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+    if (UO->getOpcode() == UO_AddrOf) {
+      return getAddressedRef(UO->getSubExpr());
+    }
+  }
+  return nullptr;
+}
+
+void checkTrivallyCopyable(
+    QualType QT, std::map<SourceLocation, std::string> &ReportLocations) {
+  const auto &Ctx = DpctGlobalInfo::getContext();
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:5030]: For replacement in ReportLocations, just insert replacements in this function.

File: clang/lib/DPCT/Utility.cpp [hunk 31]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+    while (DC) {
+      if (const NamespaceDecl *ND = dyn_cast<NamespaceDecl>(DC)) {
+        NamespaceStr = ND->getNameAsString() + "::" + NamespaceStr;
+        NS = ND;
+        DC = ND->getDeclContext();
+      } else if (isa<TranslationUnitDecl>(DC)) {
+        break;
+      } else {
+        NS = nullptr;
+        break;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]:
```suggestion
        return;
```

File: clang/lib/DPCT/Utility.cpp [hunk 32]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+      }
+    }
+    if (NS) {
+      if (NS->getEndLoc().isMacroID())
+        return;
+      InsertLoc = getInsertLocation(NS->getEndLoc(), tok::TokenKind::r_brace);
+    } else {
+      return;
+    }
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:5116]:
```suggestion
    }
```

File: clang/lib/DPCT/Utility.cpp [hunk 33]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+  if (D->getDeclContext()->getDeclKind() == Decl::Kind::TranslationUnit) {
+    if (D->getEndLoc().isMacroID())
+      return;
+    InsertLoc = getInsertLocation(D->getEndLoc(), tok::TokenKind::semi);
+  } else {
+    const DeclContext *DC = D->getDeclContext();
+    const NamespaceDecl *NS = nullptr;
+    while (DC) {
+      if (const NamespaceDecl *ND = dyn_cast<NamespaceDecl>(DC)) {
+        NamespaceStr = ND->getNameAsString() + "::" + NamespaceStr;
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: make getInsertLocation search semi by default, if goes into here, search }

File: clang/lib/DPCT/Utility.cpp [hunk 34]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+    return getAddressedRef(Cast->getSubExprAsWritten());
+  } else if (auto UO = dyn_cast<UnaryOperator>(E)) {
+    if (UO->getOpcode() == UO_AddrOf) {
+      return getAddressedRef(UO->getSubExpr());
+    }
+  }
+  return nullptr;
+}
+
+void checkTrivallyCopyable(
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: For partial specialization class, we need insert it before the specialization, not before the primary template.

File: clang/lib/DPCT/Utility.cpp [hunk 35]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+  } else if (const auto *RD = dyn_cast<RecordDecl>(D)) {
+    Repl += "template <>";
+    Repl += getNL();
+    Repl += "struct sycl::is_device_copyable<";
+    Repl += NamespaceStr;
+    Repl += RD->getNameAsString();
+    Repl += "> : std::true_type {};";
+  }
+
+  if (Rule) {
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp:5148]: 5148~5165 can be a function

File: clang/lib/DPCT/Utility.cpp [hunk 36]
```diff
@@ -4953,5 +4956,249 @@ int isArgumentInitialized(
...
+    case Type::TypeClass::Record:
+      return QT.getTypePtr()->getAs<RecordType>()->getDecl();
+    case Type::TypeClass::SubstTemplateTypeParm:
+      return getTypeDecl(QT.getTypePtr()
+                             ->castAs<SubstTemplateTypeParmType>()
+                             ->getReplacementType());
+    default:
+      return nullptr;
+    }
+  };
```
Comment (zhiweij1) [inline clang/lib/DPCT/Utility.cpp]: try to convert to a while

File: clang/lib/DPCT/Utility.cpp (hunk 34 above)
Comment (zhiweij1) [reply clang/lib/DPCT/Utility.cpp]: Test already covered.

**Review by ziranzha — APPROVED** [review]

File: clang/lib/DPCT/Diagnostics.inc [hunk 37]
```diff
@@ -282,6 +282,10 @@ DEF_WARNING(MOVE_TYPE_DEFINITION_KERNEL_FUNC, 1126, MEDIUM_LEVEL, "The type \"%0
 DEF_COMMENT(MOVE_TYPE_DEFINITION_KERNEL_FUNC, 1126, MEDIUM_LEVEL, "The type \"{0}\" defined in function \"{1}\" is used as the parameter type in all functions in the call path from the sycl::handler::parallel_for() to the function \"{1}\". You may need to adjust the type definition location.")
 DEF_WARNING(DEVICE_GLOBAL_INIT, 1127, LOW_LEVEL, "The constant compile-time initialization for device_globals is supported when compiling with C++20. You may need to adjust the compile commands.")
 DEF_COMMENT(DEVICE_GLOBAL_INIT, 1127, LOW_LEVEL, "The constant compile-time initialization for device_globals is supported when compiling with C++20. You may need to adjust the compile commands.")
+DEF_WARNING(NOT_DEVICE_COPYABLE, 1128, MEDIUM_LEVEL, "This type is used in kernel but it is not device copyable. Below part(s) is(are) not meet the requirements: %0. Please rewrite the code.")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Diagnostics.inc]:
```suggestion
DEF_WARNING(NOT_DEVICE_COPYABLE, 1128, MEDIUM_LEVEL, "The type %0 is not device copyable for %1. It is used in SYCL kernel, please rewrite the code.")
```

File: clang/lib/DPCT/Diagnostics.inc [hunk 38]
```diff
@@ -282,6 +282,10 @@ DEF_WARNING(MOVE_TYPE_DEFINITION_KERNEL_FUNC, 1126, MEDIUM_LEVEL, "The type \"%0
...
+DEF_COMMENT(NOT_DEVICE_COPYABLE, 1128, MEDIUM_LEVEL, "This type is used in kernel but it is not device copyable. Below part(s) is(are) not meet the requirements: {0}. Please rewrite the code.")
+DEF_WARNING(NOT_DEVICE_COPYABLE_ADD_SPECIALIZATION, 1129, MEDIUM_LEVEL, "The type \"%0\" is used in kernel but it is not device copyable. The tool has added the sycl::is_device_copyable specialization for this type.")
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Diagnostics.inc]:
```suggestion
DEF_WARNING(NOT_DEVICE_COPYABLE_ADD_SPECIALIZATION, 1129, MEDIUM_LEVEL, "The type \"%0\" is used in SYCL kernel but it is not device copyable.  The sycl::is_device_copyable specialization has been added for the type, please review the code.")
```

File: clang/lib/DPCT/Utility.cpp [hunk 39]
```diff
@@ -5002,5 +5005,238 @@ int isArgumentInitialized(
...
+  } else {
+    auto FileNameAndOffset = DpctGlobalInfo::getLocInfo(InsertLoc);
+    DpctGlobalInfo::getInstance().addReplacement(
+        std::make_shared<ExtReplacement>(FileNameAndOffset.first,
+                                         FileNameAndOffset.second, 0, Repl,
+                                         nullptr));
+  }
+}
+
+bool isDeviceCopyable(QualType Type, clang::dpct::MigrationRule *Rule) {
```
Comment (zhimingwang36) [inline clang/lib/DPCT/Utility.cpp:5221]: Please add comment: basic idea to judge is device copyable.

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**